README.md
Haskell FFI Bindings to NVVM
============================

[![Travis build status](https://img.shields.io/travis/tmcdonell/nvvm/master.svg?label=linux)](https://travis-ci.org/tmcdonell/nvvm)
[![AppVeyor build status](https://img.shields.io/appveyor/ci/tmcdonell/nvvm/master.svg?label=windows)](https://ci.appveyor.com/project/tmcdonell/nvvm)
[![Stackage LTS](https://stackage.org/package/nvvm/badge/lts)](https://stackage.org/lts/package/nvvm)
[![Stackage Nightly](https://stackage.org/package/nvvm/badge/nightly)](https://stackage.org/nightly/package/nvvm)
[![Hackage](https://img.shields.io/hackage/v/nvvm.svg)](https://hackage.haskell.org/package/nvvm)

The NVVM library compiles [NVVM IR][nvvm-ir-spec] (a subset of LLVM IR) into PTX code which can
then be executed on NVIDIA GPUs.

In contrast to the standard [NVPTX][nvptx-spec] target built in to the LLVM
toolchain, NVVM includes a set of proprietary optimisations which are otherwise
only available by compiling CUDA code with the `nvcc` compiler. On the other
hand, the version of LLVM that NVVM is internally based on typically lags the
public release by several generations (years), so these secret optimisations may
or may not be worthwhile to your application.

The resulting PTX code can be loaded onto the GPU and executed using the [cuda
package][hs-cuda].

The NVVM library is a compiler component available a part of the CUDA toolkit:

  <https://developer.nvidia.com/cuda-toolkit>

The configure step will look for your CUDA installation in the standard places,
and if the `nvcc` compiler is found in your `PATH`, relative to that.


[nvptx-spec]:     http://llvm.org/docs/NVPTXUsage.html
[nvvm-ir-spec]:   http://docs.nvidia.com/cuda/nvvm-ir-spec/index.html
[hs-cuda]:        https://github.com/tmcdonell/cuda



CHANGELOG.md
# Revision history for nvvm

Notable changes to the project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/).

## [0.10.0.1] - 2023-08-15
### Fixed
  * Build fixes for GHC 9.2 .. 9.6

## [0.10.0.0] - 2020-08-26
### Added
  * Support for Cabal-3
  * `nvvmDeviceLibraryPath`

## [0.9.0.0] - 2018-10-02
### Fixed
  * Build fix for ghc-8.6

### Added
  * `addModuleLazy` and `addModuleLazyFromPtr` from CUDA-10.0

### Changed
  * Replace uses of `String` with `ShortByteString`

## [0.8.0.3] - 2018-03-12
### Fixed
  * Build fix for Cabal-2.2 (ghc-8.4)

## [0.8.0.2] - 2018-01-05
### Fixed
  * Fix profiling build

## [0.8.0.1] - 2017-11-15
### Fixed
  * Fix nvvm library path on windows ([#2])

## [0.8.0.0] - 2017-08-24
### Changed
  * Build setup improvements

## [0.7.5.2] - 2017-04-10
### Added
  * Add support for older c2hs versions

## [0.7.5.1] - 2016-11-08
### Added
  * Add support for Cabal-1.22

## [0.7.5.0] - 2016-10-08
  * First version. Released on an unsuspecting world.


[0.10.0.1]:     https://github.com/tmcdonell/nvvm/compare/v0.10.0.0..v0.10.0.1
[0.10.0.0]:     https://github.com/tmcdonell/nvvm/compare/v0.9.0.0...v0.10.0.0
[0.9.0.0]:      https://github.com/tmcdonell/nvvm/compare/v0.8.0.3...v0.9.0.0
[0.8.0.3]:      https://github.com/tmcdonell/nvvm/compare/v0.8.0.2...v0.8.0.3
[0.8.0.2]:      https://github.com/tmcdonell/nvvm/compare/v0.8.0.1...v0.8.0.2
[0.8.0.1]:      https://github.com/tmcdonell/nvvm/compare/v0.8.0.0...v0.8.0.1
[0.8.0.0]:      https://github.com/tmcdonell/nvvm/compare/v0.7.5.2...v0.8.0.0
[0.7.5.2]:      https://github.com/tmcdonell/nvvm/compare/0.7.5.1...v0.7.5.2
[0.7.5.1]:      https://github.com/tmcdonell/nvvm/compare/0.7.5.0...0.7.5.1
[0.7.5.0]:      https://github.com/tmcdonell/nvvm/compare/953f6c0b99b8d667a8e261722a8daeeaba162435...0.7.5.0

[#2]:           https://github.com/tmcdonell/nvvm/pull/2



Foreign/NVVM.hs
--------------------------------------------------------------------------------
-- |
-- Module    : Foreign.NVVM
-- Copyright : [2016..2023] Trevor L. McDonell
-- License   : BSD
--
-- This module defines an interface to the /libNVVM/ library provided by NVIDIA as
-- part of the CUDA toolkit. It compiles NVVM IR, a compiler intermediate
-- representation based on LLVM IR, into PTX code suitable for execution on
-- NVIDIA GPUs. NVVM IR is a subset of LLVM IR, with a set of rules,
-- restrictions, conventions, and intrinsic functions.
--
-- NVIDIA's own 'nvcc' compiler uses NVVM IR and /libNVVM/ internally as part of
-- the CUDA C compilation process. In contrast to the (open-source) NVPTX target
-- included with the standard LLVM toolchain (which also compiles NVVM IR into
-- PTX code), /libNVVM/ includes a set of proprietary optimisation passes, which
-- /may/ result in faster GPU code. More information on NVVM IR can be found
-- here:
--
-- <http://docs.nvidia.com/cuda/nvvm-ir-spec/index.html>
--
-- The following is a short tutorial on using this library. The steps can be
-- copied into a file, or run directly in @ghci@, in which case @ghci@ should be
-- launched with the option @-fno-ghci-sandbox@. This is because CUDA maintains
-- CPU-local state, so operations must be run from a bound thread.
--
-- Note that the focus of this library is the generation of executable PTX code
-- from NVVM IR, so we will additionally need to use the 'cuda' package to
-- control and execute the compiled program. In this tutorial we will go over
-- those steps quickly, but see the 'cuda' package for more information and
-- a similar tutorial:
--
-- <https://hackage.haskell.org/package/cuda>
--
--
-- [/Initialise the CUDA environment/]
--
-- Before any operation can be performed, we must initialise the CUDA Driver
-- API.
--
-- >>> import Foreign.CUDA.Driver as CUDA
-- >>> CUDA.initialise []
--
-- Select a GPU and create an execution context for that device. Each available
-- device is given a unique numeric identifier (beginning at zero). For this
-- example we just select the first device (the default).
--
-- >>> dev0 <- CUDA.device 0
-- >>> prp0 <- CUDA.props dev0
-- >>> ctx0 <- CUDA.create dev0 []
--
-- Remember that once the context is no longer needed, it should be
-- 'Foreign.CUDA.Driver.Context.Base.destroy'ed in order to free up any
-- resources that were allocated into it.
--
-- [/Compiling kernels with NVVM/]
--
-- For this example we will step through executing the equivalent of the
-- following Haskell function, which element-wise adds the elements of two
-- arrays:
--
-- >>> vecAdd xs ys = zipWith (+) xs ys
--
-- The following NVVM IR implements this for the GPU. Note that this example is
-- written using NVVM IR version 1.2 syntax (corresponding to CUDA toolkit 7.5),
-- which is based on LLVM IR version 3.4. The human readable representation of
-- LLVM IR (and by extension NVVM IR) is notorious for changing between
-- releases, whereas the bitcode representation is somewhat more stable. You may
-- wish to keep this in mind for your own programs.
--
-- > target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v32:32:32-v64:64:64-v128:128:128-n16:32:64"
-- > target triple = "nvptx64-nvidia-cuda"
-- >
-- > define void @vecAdd(float* %A, float* %B, float* %C) {
-- > entry:
-- >   ; What is my ID?
-- >   %id = tail call i32 @llvm.nvvm.read.ptx.sreg.tid.x() readnone nounwind
-- >
-- >   ; Compute pointers into A, B, and C
-- >   %ptrA = getelementptr float* %A, i32 %id
-- >   %ptrB = getelementptr float* %B, i32 %id
-- >   %ptrC = getelementptr float* %C, i32 %id
-- >
-- >   ; Read A, B
-- >   %valA = load float* %ptrA, align 4
-- >   %valB = load float* %ptrB, align 4
-- >
-- >   ; Compute C = A + B
-- >   %valC = fadd float %valA, %valB
-- >
-- >   ; Store back to C
-- >   store float %valC, float* %ptrC, align 4
-- >
-- >   ret void
-- > }
-- >
-- > ; Intrinsic to read threadIdx.x
-- > declare i32 @llvm.nvvm.read.ptx.sreg.tid.x() readnone nounwind
-- >
-- > !nvvm.annotations = !{!0}
-- > !0 = metadata !{void (float*, float*, float*)* @vecAdd, metadata !"kernel", i64 1}
--
-- For reference, in CUDA this kernel would have been written as:
--
-- > extern "C" __global__ void vecAdd(float *xs, float* ys, float *zs)
-- > {
-- >     int ix = threadIdx.x;
-- >
-- >     zs[ix] = xs[ix] + ys[ix];
-- > }
--
-- The NVVM IR can be stored directly in the program as a @[Byte]String@, but
-- here I will assume that it is saved to a file @vector_add.ll@:
--
-- >>> import Data.ByteString as B
-- >>> ll <- B.readFile "vector_add.ll"
--
-- Now we can use NVVM to compile this into PTX code:
--
-- >>> import Foreign.NVVM as NVVM
-- >>> ptx <- NVVM.compileModule "vecAdd" ll [ NVVM.Target (CUDA.computeCapability prp0) ]
--
-- Notice that we asked NVVM to specialise the generated PTX code for our
-- current device. By default the code will be compiled for compute capability
-- 2.0 (the earliest supported target).
--
-- The generated PTX code can then be loaded into the current CUDA execution
-- context, from which we can extract a reference to the GPU kernel that we will
-- later execute.
--
-- >>> mdl    <- CUDA.loadData (NVVM.compileResult ptx)
-- >>> vecAdd <- CUDA.getFun mdl "vecAdd"
--
-- After we are finished with the module, it is a good idea to
-- 'Foreign.CUDA.Driver.Module.Base.unload' it in order to free any resources it
-- used.
--
-- [/Executing the kernel/]
--
-- Executing the 'vecAdd' kernel now proceeds exactly like executing any other
-- kernel function using the CUDA Driver API. The following is a quick overview;
-- see the tutorial in the 'cuda' package for more information.
--
-- First, generate some data and copy it to the device. We also allocate an
-- (uninitialised) array on the device to store the results.
--
-- >>> let xs = [1..256]   :: [Float]
-- >>> let ys = [2,4..512] :: [Float]
-- >>> xs_dev <- CUDA.newListArray xs
-- >>> ys_dev <- CUDA.newListArray ys
-- >>> zs_dev <- CUDA.mallocArray 256 :: IO (CUDA.DevicePtr Float)
--
-- For this simple kernel we execute it using a single (one dimensional) thread
-- block, with one thread computing each element of the output.
--
-- >>> CUDA.launchKernel vecAdd (1,1,1) (256,1,1) 0 Nothing [CUDA.VArg xs_dev, CUDA.VArg ys_dev, CUDA.VArg zs_dev]
--
-- Finally, we can copy the results back to the host, and deallocate the arrays
-- from the GPU.
--
-- >>> zs <- CUDA.peekListArray 256 zs_dev
-- >>> CUDA.free xs_dev
-- >>> CUDA.free ys_dev
-- >>> CUDA.free zs_dev
--
--
-- [/Next steps/]
--
-- The library also provides functions for compiling several NVVM IR sources
-- into a single module. In particular this is useful when linking against
-- @libdevice@, a standard library of functions in NVVM IR which implement, for
-- example, math primitives and bitwise operations. More information on
-- @libdevice@ can be found here:
--
-- <http://docs.nvidia.com/cuda/libdevice-users-guide/index.html>
--
--------------------------------------------------------------------------------

module Foreign.NVVM (

  module Foreign.NVVM.Compile,
  module Foreign.NVVM.Error,
  module Foreign.NVVM.Info,

) where

import Foreign.NVVM.Compile
import Foreign.NVVM.Error
import Foreign.NVVM.Info



Setup.hs
{-# LANGUAGE CPP                  #-}
{-# LANGUAGE FlexibleInstances    #-}
{-# LANGUAGE QuasiQuotes          #-}
{-# LANGUAGE TemplateHaskell      #-}
{-# LANGUAGE OverloadedStrings    #-}
{-# LANGUAGE ExtendedDefaultRules #-}

-- This macro is only available when compiling with GHC-8
#ifndef MIN_VERSION_Cabal
#define MIN_VERSION_Cabal(major1,major2,minor) 0
#endif

import Distribution.PackageDescription
import Distribution.Simple
import Distribution.Simple.BuildPaths
import Distribution.Simple.Command
import Distribution.Simple.LocalBuildInfo
import Distribution.Simple.PreProcess
import Distribution.Simple.Program
import Distribution.Simple.Program.Db
import Distribution.Simple.Program.Find
import Distribution.Simple.Setup
import Distribution.Simple.Utils
import Distribution.System
import Distribution.Verbosity

#if MIN_VERSION_Cabal(1,25,0)
import Distribution.PackageDescription.PrettyPrint
import Distribution.Version
#endif
#if MIN_VERSION_Cabal(2,2,0)
import Distribution.PackageDescription.Parsec
#else
import Distribution.PackageDescription.Parse
#endif
#if MIN_VERSION_Cabal(3,8,0)
import Distribution.Simple.PackageDescription
#endif

import Foreign.CUDA.Path
import Control.Applicative
import Control.Exception
import Control.Monad
import Data.List
import System.Directory
import System.FilePath
import Text.Printf
import Data.Char
import Prelude

import Language.Haskell.TH

import qualified Data.Text as T

default (String)

-- Configuration
-- -------------

customBuildInfoFilePath :: FilePath
customBuildInfoFilePath = "nvvm" <.> "buildinfo"

generatedBuildInfoFilePath :: FilePath
generatedBuildInfoFilePath = customBuildInfoFilePath <.> "generated"

-- Location of the NVVM library relative to the base CUDA installation
--
nvvmPath :: Platform -> FilePath
nvvmPath _ = cudaInstallPath </> "nvvm"

nvvmIncludePath :: Platform -> FilePath
nvvmIncludePath platform = nvvmPath platform </> "include"

nvvmLibraryPath :: Platform -> FilePath
nvvmLibraryPath platform@(Platform arch os) = nvvmPath platform </> libpath
  where
    libpath =
      case (os, arch) of
        (Windows, I386)   -> "lib" </> "Win32"
        (Windows, X86_64) -> "lib" </> "x64"
        (OSX,     _)      -> "lib"    -- MacOS does not distinguish 32- vs. 64-bit paths
        (_,       X86_64) -> "lib64"  -- treat all others similarly
        _                 -> "lib"

nvvmLibrary :: Platform -> [String]
nvvmLibrary _ = ["nvvm"]


-- Build setup
-- -----------

main :: IO ()
main = defaultMainWithHooks customHooks
  where
    readHook get_verbosity args flags = do
        getHookedBuildInfo (fromFlag (get_verbosity flags))

    preprocessors = hookedPreProcessors simpleUserHooks

    -- Our readHook implementation uses our getHookedBuildInfo. We can't rely on
    -- cabal's autoconfUserHooks since they don't handle user overwrites to
    -- buildinfo like we do.
    --
    customHooks =
      simpleUserHooks
        { preBuild            = preBuildHook -- not using 'readHook' here because 'build' takes extra args
        , preClean            = readHook cleanVerbosity
        , preCopy             = readHook copyVerbosity
        , preInst             = readHook installVerbosity
        , preHscolour         = readHook hscolourVerbosity
        , preHaddock          = readHook haddockVerbosity
        , preReg              = readHook regVerbosity
        , preUnreg            = readHook regVerbosity
        , postConf            = postConfHook
        , postBuild           = postBuildHook
        , hookedPreProcessors = ("chs", pp_c2hs) : filter (\x -> fst x /= "chs") preprocessors
        }

    -- The hook just loads the HookedBuildInfo generated by postConfHook, unless
    -- there is user-provided info that overwrites it.
    --
    preBuildHook :: Args -> BuildFlags -> IO HookedBuildInfo
    preBuildHook _ flags = getHookedBuildInfo $ fromFlag $ buildVerbosity flags

    -- The hook scans system in search for CUDA Toolkit. If the toolkit is not
    -- found, an error is raised. Otherwise the toolkit location is used to
    -- create a `nvvm.buildinfo.generated` file with all the resulting flags.
    postConfHook :: Args -> ConfigFlags -> PackageDescription -> LocalBuildInfo -> IO ()
    postConfHook args flags pkg_descr lbi = do
      let
          verbosity       = fromFlagOrDefault normal (configVerbosity flags)
          profile         = fromFlagOrDefault False  (configProfLib flags)
          currentPlatform = hostPlatform lbi
          compilerId_     = compilerId (compiler lbi)
      --
      generateAndStoreBuildInfo verbosity profile currentPlatform compilerId_ generatedBuildInfoFilePath
      validateLinker verbosity currentPlatform $ withPrograms lbi
      --
      actualBuildInfoToUse <- getHookedBuildInfo verbosity
      let pkg_descr' = updatePackageDescription actualBuildInfoToUse pkg_descr
      postConf simpleUserHooks args flags pkg_descr' lbi

    -- This hook fixes the embedded LC_RPATHs in the generated .dylib on OSX.
    postBuildHook :: Args -> BuildFlags -> PackageDescription -> LocalBuildInfo -> IO ()
    postBuildHook _ flags pkg_descr lbi = do
      let
          verbosity           = fromFlag (buildVerbosity flags)
          platform            = hostPlatform lbi
          cid                 = compilerId (compiler lbi)
          uid                 =
#if MIN_VERSION_Cabal(1,25,0)
            localUnitId lbi
#else
            $( case withinRange cabalVersion (orLaterVersion (Version [1,24] [])) of
                 True  -> return ( AppE (VarE (mkName "localUnitId")) (VarE (mkName "lbi")) )
                 False -> return ( AppE (VarE (mkName "head"))
                                        (AppE (VarE (mkName ("componentLibraries")))
                                        (AppE (AppE (VarE (mkName "getComponentLocalBuildInfo")) (VarE (mkName "lbi"))) (ConE (mkName "CLibName")))) )

                 -- True  -> [| localUnitId lbi |]
                 -- False -> [| head (componentLibraries (getComponentLocalBuildInfo lbi CLibName)) |]
             )
#endif
#if MIN_VERSION_Cabal(2,3,0)
          sharedLib           = buildDir lbi </> mkSharedLibName platform cid uid
#else
          sharedLib           = buildDir lbi </> mkSharedLibName          cid uid
#endif
          Just extraLibDirs'  = extraLibDirs . libBuildInfo <$> library pkg_descr
      --
      updateLibraryRPATHs verbosity platform sharedLib extraLibDirs'


-- It seems that the GHC and/or Cabal developers don't quite understand how
-- dynamic linking works on OSX. Even though we have specified
-- '-optl-Wl,-rpath,...' as part of the configuration, this (sometimes?) gets
-- filtered out somewhere, and the resulting .dylib that is generated does not
-- have this path embedded as an LC_RPATH. The result is that the NVVM library
-- will not be found, resulting in a link-time error.
--
-- On *nix (and versions of OSX previous to El Capitan 10.11), we could use
-- [DY]LD_LIBRARY_PATH to specify where to resolve @rpath locations, but that is
-- no longer an option on OSX due to System Integrity Protection.
--
-- An alternate argument is that the CUDA installer should have updated the
-- install name (LC_ID_DYLIB) of its dynamic libraries to include the full
-- absolute path, rather than relying on @rpath in the first place, which is
-- what Apple's system libraries do for example.
--
updateLibraryRPATHs :: Verbosity -> Platform -> FilePath -> [FilePath] -> IO ()
updateLibraryRPATHs verbosity pf@(Platform _ os) sharedLib extraLibDirs' =
  when (os == OSX) $ do
    exists <- doesFileExist sharedLib
    unless exists $ die' verbosity $ printf "Unexpected failure: library does not exist: %s" sharedLib
    --
    rpaths <- getLibraryRPATHs verbosity pf sharedLib
    mint   <- findProgram verbosity "install_name_tool"
    case mint of
      Nothing                -> notice verbosity $ "Could not locate 'install_name_tool' in order to update LC_RPATH entries. This is likely to cause problems later on."
      Just install_name_tool ->
        forM_ (extraLibDirs' \\ rpaths) $ \libDir ->
          runProgramInvocation verbosity $ simpleProgramInvocation install_name_tool ["-add_rpath", libDir, sharedLib]

parseRPATHs :: String -> [FilePath]
parseRPATHs = loop [] . T.lines . T.pack
  where
    loop acc xs =
      case break (T.isPrefixOf "Load command") xs of
        (_, []) -> acc
        (_, _ : cmd : rest)
          | T.stripPrefix "cmd " (T.stripStart cmd) == Just "LC_RPATH"
            -> let l:ls = dropWhile (not . T.isPrefixOf "path " . T.stripStart) rest
                   Just (path, _) = T.breakOnEnd " (offset" <$> T.stripPrefix "path " (T.stripStart l)
               in loop (T.unpack (T.dropEnd 8 path) : acc) ls
          | otherwise -> loop acc rest
    

getLibraryRPATHs :: Verbosity -> Platform -> FilePath -> IO [FilePath]
getLibraryRPATHs verbosity (Platform _ os) sharedLib =
  if os /= OSX
  then return []
  else do
    exists <- doesFileExist sharedLib
    unless exists $ die' verbosity $ printf "Unexpected failure: library does not exist: %s" sharedLib
    mint   <- findProgram verbosity "otool"
    case mint of
      Nothing    -> do
        notice verbosity $ "Could not locate 'otool' in order to get LC_RPATH entries. This is likely to cause problems later on."
        return []
      Just otool ->
        parseRPATHs <$> getProgramInvocationOutput verbosity (simpleProgramInvocation otool ["-l", sharedLib])


-- Reads user-provided `nvvm.buildinfo` if present, otherwise loads `nvvm.buildinfo.generated`
-- Outputs message informing about the other possibility.
-- Calls die when neither of the files is available.
-- (generated one should be always present, as it is created in the post-conf step)
--
getHookedBuildInfo :: Verbosity -> IO HookedBuildInfo
getHookedBuildInfo verbosity = do
  customBuildInfoExists <- doesFileExist customBuildInfoFilePath
  if customBuildInfoExists
    then do
      notice verbosity $ printf "The user-provided buildinfo from file '%s' will be used. To use default settings, delete this file." customBuildInfoFilePath
      readHookedBuildInfo verbosity customBuildInfoFilePath
    else do
      generatedBuildInfoExists <- doesFileExist generatedBuildInfoFilePath
      if generatedBuildInfoExists
        then do
          notice verbosity $ printf "Using build information from '%s'" generatedBuildInfoFilePath
          notice verbosity $ printf "Provide a '%s' file to override this behaviour" customBuildInfoFilePath
          readHookedBuildInfo verbosity generatedBuildInfoFilePath
        else
          die' verbosity $ printf "Unexpected failure: neither the default '%s' nor custom '%s' exist" generatedBuildInfoFilePath customBuildInfoFilePath


-- Runs CUDA detection procedure and stores .buildinfo to a file.
--
generateAndStoreBuildInfo :: Verbosity -> Bool -> Platform -> CompilerId -> FilePath -> IO ()
generateAndStoreBuildInfo verbosity profile platform (CompilerId _ghcFlavor ghcVersion) path =
  storeHookedBuildInfo verbosity path =<< libraryBuildInfo profile platform ghcVersion

findProgram :: Verbosity -> FilePath -> IO (Maybe FilePath)
findProgram verbosity prog =
  findProgram_helper (findProgramOnSearchPath verbosity defaultProgramSearchPath prog)

class FindProgram f where
  findProgram_helper :: f -> IO (Maybe FilePath)

instance FindProgram (IO (Maybe (FilePath, [FilePath]))) where
  findProgram_helper f = fmap fst `fmap` f

instance FindProgram (IO (Maybe FilePath)) where
  findProgram_helper f = f


-- Generates build info with flags needed for CUDA Toolkit to be properly
-- visible to underlying build tools.
--
libraryBuildInfo :: Bool -> Platform -> Version -> IO HookedBuildInfo
libraryBuildInfo profile platform@(Platform arch os) ghcVersion = do
  let
      -- XXX
      libraryPaths      = [nvvmLibraryPath platform]
      includePaths      = [nvvmIncludePath platform, cudaIncludePath]

      -- options for GHC
      extraLibDirs'     = libraryPaths
      ccOptions'        = map ("-I"++) includePaths
      ldOptions'        = map ("-L"++) extraLibDirs'
      ghcOptions        = map ("-optc"++) ccOptions'
                       ++ map ("-optl"++) ldOptions'
                       ++ if os /= Windows && not profile
                            then map ("-optl-Wl,-rpath,"++) extraLibDirs'
                            else []
      extraLibs'        = nvvmLibrary platform

      -- options for C2HS
      archFlag          = case arch of
                            I386   -> "-m32"
                            X86_64 -> "-m64"
                            _      -> ""
      emptyCase         = ["-DUSE_EMPTY_CASE" | versionBranch ghcVersion >= [7,8]]
      blocksExtension   = ["-U__BLOCKS__" | os == OSX ]
      c2hsOptions       = unwords $ map ("--cppopts="++) ("-E" : archFlag : emptyCase ++ blocksExtension)
      c2hsExtraOptions  = ("x-extra-c2hs-options", c2hsOptions)

      addSystemSpecificOptions :: BuildInfo -> IO BuildInfo
      addSystemSpecificOptions bi =
        case os of
          -- In the CUDA package this is used to populate the extraGHCiLibs
          -- field with the mangled .dll names. I'm not sure what those are for
          -- this library, so left out for the time being.
          Windows -> return bi
          OSX     -> return bi
          _       -> return bi

  buildInfo' <- addSystemSpecificOptions $ emptyBuildInfo
    { ccOptions      = ccOptions'
    , ldOptions      = ldOptions'
    , extraLibs      = extraLibs'
    , extraLibDirs   = extraLibDirs'
#if MIN_VERSION_Cabal(3,0,0)
    , options        = PerCompilerFlavor (if os /= Windows then ghcOptions else []) []
#else
    , options        = [(GHC, ghcOptions) | os /= Windows]
#endif
    , customFieldsBI = [c2hsExtraOptions]
    }

  return (Just buildInfo', [])


storeHookedBuildInfo :: Verbosity -> FilePath -> HookedBuildInfo -> IO ()
storeHookedBuildInfo verbosity path hbi = do
    notice verbosity $ "Storing parameters to " ++ path
    writeHookedBuildInfo path hbi


-- On Windows platform the binutils linker targeting x64 is bugged and cannot
-- properly link with import libraries generated by MS compiler (like the CUDA ones).
-- The programs would correctly compile and crash as soon as the first FFI call is made.
--
-- Therefore we fail configure process if the linker is too old and provide user
-- with guidelines on how to fix the problem.
--
validateLinker :: Verbosity -> Platform -> ProgramDb -> IO ()
validateLinker verbosity (Platform X86_64 Windows) db = do
  let say msg = printf "%s. If generated executables crash when making calls to NVVM please see: %s" msg windowsHelpPage
  --
  maybeLdPath <- getRealLdPath verbosity db
  case maybeLdPath of
    Nothing     -> warn verbosity $ say "Cannot find ld.exe to check if it is new enough"
    Just ldPath -> do
      debug verbosity $ "Checking if ld.exe at " ++ ldPath ++ " is new enough"
      maybeVersion <- getLdVersion verbosity ldPath
      case maybeVersion of
        Nothing        -> warn verbosity $ say "Unknown ld.exe version"
        Just ldVersion -> do
          debug verbosity $ "Found ld.exe version: " ++ show ldVersion
          when (ldVersion < [2,25,1]) $ die' verbosity (windowsLinkerBugMsg ldPath)
validateLinker _ _ _ = return () -- The linker bug is present only on Win64 platform


-- On Windows GHC package comes with two copies of ld.exe.
--
-- ProgramDb knows about the first one: ghcpath\mingw\bin\ld.exe
-- This function returns the other one: ghcpath\mingw\x86_64-w64-mingw32\bin\ld.exe
--
-- The second one is the one that does actual linking and code generation.
-- See: https://github.com/tmcdonell/cuda/issues/31#issuecomment-149181376
--
-- The function is meant to be used only on 64-bit GHC distributions.
--
getRealLdPath :: Verbosity -> ProgramDb -> IO (Maybe FilePath)
getRealLdPath verbosity programDb =
  -- This should ideally work `programFindVersion ldProgram` but for some reason
  -- it does not. The issue should be investigated at some time.
  case lookupProgram ghcProgram programDb of
    Nothing            -> return Nothing
    Just configuredGhc -> do
      let ghcPath        = locationPath $ programLocation configuredGhc
          presumedLdPath = (takeDirectory . takeDirectory) ghcPath </> "mingw" </> "x86_64-w64-mingw32" </> "bin" </> "ld.exe"
      info verbosity $ "Presuming ld location" ++ presumedLdPath
      presumedLdExists <- doesFileExist presumedLdPath
      return $ if presumedLdExists
                 then Just presumedLdPath
                 else Nothing

-- Tries to obtain the version `ld`. Throws an exception on failure.
--
getLdVersion :: Verbosity -> FilePath -> IO (Maybe [Int])
getLdVersion verbosity ldPath = do
  -- Examples of version string format:
  --  * GNU ld (GNU Binutils) 2.25.1
  --  * GNU ld (GNU Binutils) 2.20.51.20100613
  --
  ldVersionString <- getProgramInvocationOutput normal (simpleProgramInvocation ldPath ["-v"])

  let versionText   = last $ words ldVersionString -- takes e.g. "2.25.1"
      versionParts  = splitOn (== '.') versionText
      versionParsed = Just $ map read versionParts

  -- last and read above may throw and message would be not understandable for user,
  -- so we'll intercept exception and rethrow it with more useful message.
  let handleError :: SomeException -> IO (Maybe [Int])
      handleError e = do
          warn verbosity $ printf "cannot parse ld version string: '%s'. Parsing exception: %s" ldVersionString (show e)
          return Nothing

  evaluate versionParsed `catch` handleError

splitOn :: (Char -> Bool) -> String -> [String]
splitOn p s =
  case dropWhile p s of
    [] -> []
    ss -> let (w,s') = break p ss in w : splitOn p s'


windowsHelpPage :: String
windowsHelpPage = "https://github.com/tmcdonell/nvvm/blob/master/WINDOWS.md"

windowsLinkerBugMsg :: FilePath -> String
windowsLinkerBugMsg ldPath = printf (unlines msg) windowsHelpPage ldPath
  where
    msg =
      [ "********************************************************************************"
      , ""
      , "The installed version of `ld.exe` has version < 2.25.1. This version has known bug on Windows x64 architecture, making it unable to correctly link programs using CUDA. The fix is available and MSys2 released fixed version of `ld.exe` as part of their binutils package (version 2.25.1)."
      , ""
      , "To fix this issue, replace the `ld.exe` in your GHC installation with the correct binary. See the following page for details:"
      , ""
      , "  %s"
      , ""
      , "The full path to the outdated `ld.exe` detected in your installation:"
      , ""
      , "> %s"
      , ""
      , "Please download a recent version of binutils `ld.exe`, from, e.g.:"
      , ""
      , "  http://repo.msys2.org/mingw/x86_64/mingw-w64-x86_64-binutils-2.25.1-1-any.pkg.tar.xz"
      , ""
      , "********************************************************************************"
      ]


-- Replicate the default C2HS preprocessor hook here, and inject a value for
-- extra-c2hs-options, if it was present in the buildinfo file
--
-- Everything below copied from Distribution.Simple.PreProcess
--
class PPC2HS f where
  pp_c2hs :: f

instance PPC2HS (BuildInfo -> LocalBuildInfo -> ComponentLocalBuildInfo -> PreProcessor) where
  pp_c2hs bi lbi _ = pp_c2hs bi lbi

instance PPC2HS (BuildInfo -> LocalBuildInfo -> PreProcessor) where
  pp_c2hs bi lbi =
    PreProcessor
      { platformIndependent = False
#if MIN_VERSION_Cabal(3,8,0)
      , ppOrdering          = unsorted
#endif
      , runPreProcessor     = \(inBaseDir, inRelativeFile)
                               (outBaseDir, outRelativeFile) verbosity ->
          runDbProgram verbosity c2hsProgram (withPrograms lbi) . filter (not . null) $
            maybe [] words (lookup "x-extra-c2hs-options" (customFieldsBI bi))
            ++ ["--include=" ++ outBaseDir]
            ++ ["--cppopts=" ++ opt | opt <- getCppOptions bi lbi]
            ++ ["--output-dir=" ++ outBaseDir,
                "--output=" ++ outRelativeFile,
                inBaseDir </> inRelativeFile]
      }

getCppOptions :: BuildInfo -> LocalBuildInfo -> [String]
getCppOptions bi lbi
    = hcDefines (compiler lbi)
   ++ ["-I" ++ dir | dir <- includeDirs bi]
   ++ [opt | opt@('-':c:_) <- ccOptions bi, c `elem` ("DIU" :: String)]

hcDefines :: Compiler -> [String]
hcDefines comp =
  case compilerFlavor comp of
    GHC  -> ["-D__GLASGOW_HASKELL__=" ++ versionInt version]
    JHC  -> ["-D__JHC__=" ++ versionInt version]
    NHC  -> ["-D__NHC__=" ++ versionInt version]
    Hugs -> ["-D__HUGS__"]
    _    -> []
  where version = compilerVersion comp

-- TODO: move this into the compiler abstraction
-- FIXME: this forces GHC's crazy 4.8.2 -> 408 convention on all the other
-- compilers. Check if that's really what they want.
versionInt :: Version -> String
versionInt v =
  case versionBranch v of
    []      -> "1"
    [n]     -> show n
    n1:n2:_ -> printf "%d%02d" n1 n2

#if MIN_VERSION_Cabal(1,25,0)
versionBranch :: Version -> [Int]
versionBranch = versionNumbers
#endif

#if !MIN_VERSION_Cabal(2,0,0)
die' :: Verbosity -> String -> IO a
die' _ = die
#endif

